src/writer.rs

```diff
@@ -0,0 +1,373 @@
+    let slope = if header.scl_slope == 0.0 { 1.0 } else { header.scl_slope };
+    if slope != 1.0 || header.scl_inter != 0.0 {
+        let slope = A::from_f32(slope).unwrap();
+        let inter = A::from_f32(header.scl_inter).unwrap();
+        for arr_data in data.axis_iter(Axis(0)) {
+            let arr_data = arr_data.sub(inter).div(slope);
+            let arr_data = arr_data.into_shape(arr_len).unwrap();
+            let arr_data = arr_data.as_slice().unwrap();
+
+            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
+            writer.write_all(&buffer).unwrap();
+        }
+    } else {
+        for arr_data in data.axis_iter(Axis(0)) {
+            // We need to own the data because of the into_shape() for `C` ordering.
+            let arr_data = arr_data.to_owned();
+            let arr_data = arr_data.into_shape(arr_len).unwrap();
+            let arr_data = arr_data.as_slice().unwrap();
+
+            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
+            writer.write_all(&buffer).unwrap();
+        }
+    }
+}
+
+fn to_bytes<T>(data: &[T], size: usize) -> &[u8] {
+    unsafe { from_raw_parts::<u8>(data.as_ptr() as *const u8, size) }
+}
+
+#[cfg(test)]
+pub mod tests {
+    extern crate tempfile;
+
+    use super::*;
+
+    use ndarray::{Array, Array2, Ix2, IxDyn, ShapeBuilder};
+    use self::tempfile::tempdir;
+
+    use {InMemNiftiObject, IntoNdArray, header::MAGIC_CODE_NIP1, object::NiftiObject};
+
+    fn get_random_path(ext: &str) -> String {
+        let dir = tempdir().unwrap();
+        let path = if ext == "" {
+            dir.into_path()
+        } else {
+            dir.into_path().join(ext)
+        };
+        path.to_str().unwrap().to_string()
+    }
+
+    pub fn generate_nifti_header(
+        dim: [u16; 8],
+        scl_slope: f32,
+        scl_inter: f32,
+        datatype: i16
+    ) -> NiftiHeader {
+        let bitpix = (mem::size_of::<f32>() * 8) as i16;
+        let magic = *MAGIC_CODE_NIP1;
+        NiftiHeader {
+            dim, datatype, bitpix, magic, scl_slope, scl_inter, ..NiftiHeader::default()
+        }
+    }
+
+    fn read_2d_image(path: &str) -> Array2<f32> {
+        let nifti_object = InMemNiftiObject::from_file(path).expect("Nifti file is unreadable.");
+        let volume = nifti_object.into_volume();
+        let dyn_data = volume.into_ndarray().unwrap();
+        dyn_data.into_dimensionality::<Ix2>().unwrap()
+    }
+
+    fn test_write_read(arr: &Array<f32, IxDyn>, path: &str) {
+        let path = get_random_path(path);
+        let mut dim = [1; 8];
+        dim[0] = arr.ndim() as u16;
+        for (i, s) in arr.shape().iter().enumerate() {
+            dim[i + 1] = *s as u16;
+        }
+        let header = generate_nifti_header(dim, 1.0, 0.0, 16);
+        write_nifti(&path, &arr, Some(&header));
+
+        let read_nifti = read_2d_image(&path);
+        assert!(read_nifti.all_close(&arr, 1e-10) == true);
```

Comment:
That `== true` is unnecessary IMO.
```suggestion
        assert!(read_nifti.all_close(&arr, 1e-10));
```